extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -680,59 +713,67 @@ public void run()
-                        true,
-                        // do not allow incremental persists to happen until all the rows from this batch
-                        // of rows are indexed
-                        false
-                    );
-
-                    if (addResult.isOk()) {
-                      // If the number of rows in the segment exceeds the threshold after adding a row,
-                      // move the segment out from the active segments of BaseAppenderatorDriver to make a new segment.
-                      if (addResult.getNumRowsInSegment() > tuningConfig.getMaxRowsPerSegment()) {
-                        if (!sequenceToUse.isCheckpointed()) {
-                          sequenceToCheckpoint = sequenceToUse;
+                      if (addResult.isOk()) {
+                        // If the number of rows in the segment exceeds the threshold after adding a row,
+                        // move the segment out from the active segments of BaseAppenderatorDriver to make a new segment.
+                        if (addResult.getNumRowsInSegment() > tuningConfig.getMaxRowsPerSegment()) {
+                          if (!sequenceToUse.isCheckpointed()) {
+                            sequenceToCheckpoint = sequenceToUse;
+                          }
                         }
+                        isPersistRequired |= addResult.isPersistRequired();
+                      } else {
+                        // Failure to allocate segment puts determinism at risk, bail out to be safe.
+                        // May want configurable behavior here at some point.
+                        // If we allow continuing, then consider blacklisting the interval for a while to avoid constant checks.
+                        throw new ISE("Could not allocate segment for row with timestamp[%s]", row.getTimestamp());
                       }
-                      isPersistRequired |= addResult.isPersistRequired();
+
+                      if (addResult.getParseException() != null) {
+                        throw addResult.getParseException();
+                      }
+                      fireDepartmentMetrics.incrementProcessed();
                     } else {
-                      // Failure to allocate segment puts determinism at risk, bail out to be safe.
-                      // May want configurable behavior here at some point.
-                      // If we allow continuing, then consider blacklisting the interval for a while to avoid constant checks.
-                      throw new ISE("Could not allocate segment for row with timestamp[%s]", row.getTimestamp());
+                      fireDepartmentMetrics.incrementThrownAway();
                     }
-
-                    fireDepartmentMetrics.incrementProcessed();
-                  } else {
-                    fireDepartmentMetrics.incrementThrownAway();
+                  }
+                  catch (ParseException e) {
```

Comment:
Looks like this `catch` clause isn't necessary and handleParseException() can be called directly like 

```
if (addResult.getParseException() != null) {
  handleParseException(e, record);
}
```

Reply:
Changed this and similar places to use handleParseException